Datastore/src/Operation.php

```diff
@@ -871,4 +961,275 @@ private function sortEntities(array $entities, array $keys)
+     * @return array
+     */
+    private function parseQuery(array $query)
+    {
+        if (isset($query['order'])) {
+            foreach ($query['order'] as &$order) {
+                $order['direction'] = $order['direction'] === 'ASCENDING'
+                    ? Direction::ASCENDING
+                    : Direction::DESCENDING;
+            }
+        }
+
+        if (isset($query['filter'])) {
+            $query['filter'] = $this->convertFilterProps($query['filter']);
+        }
+
+        if (isset($query['limit']) && !is_array($query['limit'])) {
+            $query['limit'] = [
+                'value' => $query['limit']
+            ];
+        }
+
+        return $query;
+    }
+
+    /**
+     * Parse query into API compatible values
+     *
+     * @param array $gqlQuery
+     * @return array
+     */
+    private function parseGqlQuery(array $gqlQuery)
+    {
+        if (isset($gqlQuery['namedBindings'])) {
+            foreach ($gqlQuery['namedBindings'] as $name => &$binding) {
+                if (!isset($binding['value'])) {
+                    continue;
+                }
+
+                $binding = $this->prepareQueryBinding($binding);
+            }
+        }
+
+        if (isset($gqlQuery['positionalBindings'])) {
+            foreach ($gqlQuery['positionalBindings'] as &$binding) {
+                if (!isset($binding['value'])) {
+                    continue;
+                }
+
+                $binding = $this->prepareQueryBinding($binding);
+            }
+        }
+
+        return $gqlQuery;
+    }
+
+    /**
+     * Convert Query filters to an API-compatible value.
+     *
+     * @param array $filter The input filter data
+     * @return array
+     */
+    private function convertFilterProps(array $filter)
+    {
+        if (isset($filter['propertyFilter'])) {
+            $operator = $filter['propertyFilter']['op'];
+
+            try {
+                if (is_int($operator)) {
+                    // verify that the operator, given as enum value, exists.
+                    PropertyFilterOperator::name($operator);
+                } else {
+                    // convert the operator, given as a string, to a grpc value.
+                    $operator = PropertyFilterOperator::value($operator);
+                }
+            } catch (\UnexpectedValueException $e) {
+                throw new \InvalidArgumentException($e->getMessage());
+            }
+
+            $filter['propertyFilter']['op'] = $operator;
+        }
+
+        if (isset($filter['compositeFilter'])) {
+            if ($filter['compositeFilter']['op'] == 'AND') {
+                $filter['compositeFilter']['op'] = CompositeFilterOperator::PBAND;
+            } elseif ($filter['compositeFilter']['op'] == 'OR') {
+                $filter['compositeFilter']['op'] = CompositeFilterOperator::PBOR;
+            } else {
+                $filter['compositeFilter']['op'] = CompositeFilterOperator::OPERATOR_UNSPECIFIED;
+            }
```

Comment:
why are we not doing the same thing here as above and just call `CompositeFilterOperator::value()`? The way we have it here, if the `"op"` is anything but `AND` or `OR`, it gets overwritten by `OPERATOR_UNSPECIFIED`.

Reply:
This composite filter feature request came last year. The property filter was already present by then. I checked and found that backend verifies the operator value and Operator unspecified results in an INVALID_ARGUMENT api error from backend. Hence omitted the check at the client library level.

The property filter's check seems excessive at that time so I didn't follow that. 

_The code flow ideally shouldn't reach the third else part because we've exposed composite filters via the Filter [class](https://github.com/googleapis/google-cloud-php-datastore/blob/main/src/Query/Filter.php#L76-L98) which doesn't allows users to create filters with custom operator values from scratch_